src/libraries/System.Private.CoreLib/src/System/Text/CompositeFormat.cs

```diff
@@ -0,0 +1,350 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Collections.Generic;
+using System.Diagnostics;
+using System.Diagnostics.CodeAnalysis;
+using System.Runtime.CompilerServices;
+
+namespace System.Text
+{
+    /// <summary>Represents a parsed composite format string.</summary>
+    [DebuggerDisplay("{Format}")]
+    public sealed class CompositeFormat
```

Comment:
Why a class instead of a readonly struct? The vast majority use case for this will be held as either a static, or a field in a long-lived class. Readonly struct eliminates indirection overhead and allocation overhead...

Reply:
It's passed as an argument to non-inlined methods, so it will get copied on every call.  As a class, passing it around is just passing around a pointer, we don't need to be concerned about the size of the instance growing in the future (if, for example, we were to do something special to pre-parse format specifiers as well), etc.  We could also in the future choose to specialize with derived types, virtual methods, etc., if deemed valuable.  There's not really a benefit here to it being a struct, and in general unless there is a strong benefit we default to using reference types.

Reply:
Along with making it a struct, I would've used the type using 'in' parameters so no copying would take place.

I obsess over dereferences and total # of cache lines involved in operations. Making something a reference type means needing an extra dereference operation and 1 or 2 extra data cache lines to access the object. Whereas if it is a struct in a static or a struct in a larger instance, you avoid this.

Not a big deal. Just a bit of peanut butter...

Reply:
If a struct S is inlined in a larger object O, you can access S' members as offsets relative to 'this'. If the object O instead contains a reference to a heap object, then you need to load the reference out of O first in order to access the fields of the heap object. It doesn't change the code that implements this heap object, except when inlining occurs at which points it ends up being more efficient by avoiding the need to go through a reference.